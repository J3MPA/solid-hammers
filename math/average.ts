import { isRealNumber } from '../functions/assert/isRealNumber'
import { ifThis } from '../functions/assert/ifThis'
import { isFalse } from '../functions/assert/isFalse'

const throwTypeError = (val: any) => {
  throw new TypeError(`"${val}" is not a real number`)
}

const isNotARealNumber = (num: any) => isFalse(isRealNumber(num))

export const average = (nNumbers: number[]) =>
  nNumbers.reduce((acc, number, i, arr) => {
    ifThis(number, isNotARealNumber, throwTypeError)

    const nth = arr.length
    const isLast = i === nth - 1
    if (isLast) return (acc + number) / nth
    return acc + number
  }, 0)
