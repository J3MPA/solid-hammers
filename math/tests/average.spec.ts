import { average } from '../average'

describe('average', () => {
  describe('Happy path', () => {
    it('should return avg values given an array of number', () => {
      expect(average([10, 10, 10])).toBe(10)
      expect(average([20, 5, 5])).toBe(10)
      expect(average([5, 15, 10])).toBe(10)
      expect(average([65, 78, 63, 352])).toBe(139.5)
    })
  })

  describe('Sad path', () => {
    it('should trow type error if one of given numbers is not a real number', () => {
      expect(() => average([10, '10', 10] as any)).toThrowError(TypeError)
      expect(() =>
        average([10, '10', 10] as any)
      ).toThrowErrorMatchingSnapshot()
      expect(() => average([20, 5, null] as any)).toThrowError(TypeError)
      expect(() => average([20, 5, null] as any)).toThrowErrorMatchingSnapshot()
      expect(() => average([5, 15, undefined] as any)).toThrowError(TypeError)
      expect(() =>
        average([5, 15, undefined] as any)
      ).toThrowErrorMatchingSnapshot()
      expect(() => average([65, 78, 63, NaN] as any)).toThrowError(TypeError)
      expect(() =>
        average([65, 78, 63, NaN] as any)
      ).toThrowErrorMatchingSnapshot()
      expect(() => average([65, 78, 63, Infinity] as any)).toThrowError(
        TypeError
      )
      expect(() =>
        average([65, 78, 63, Infinity] as any)
      ).toThrowErrorMatchingSnapshot()
      expect(() => average([65, 78, 63, []] as any)).toThrowError(TypeError)
      expect(() =>
        average([65, 78, 63, []] as any)
      ).toThrowErrorMatchingSnapshot()
      expect(() => average([65, 78, 63, {}] as any)).toThrowError(TypeError)
      expect(() =>
        average([65, 78, 63, {}] as any)
      ).toThrowErrorMatchingSnapshot()
      expect(() => average([65, 78, 63, ''] as any)).toThrowError(TypeError)
      expect(() =>
        average([65, 78, 63, ''] as any)
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
