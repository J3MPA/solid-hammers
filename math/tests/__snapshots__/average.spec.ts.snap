// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`average Sad path should trow type error if one of given numbers is not a real number 1`] = `""10" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 2`] = `""null" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 3`] = `""undefined" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 4`] = `""NaN" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 5`] = `""Infinity" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 6`] = `""" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 7`] = `""[object Object]" is not a real number"`;

exports[`average Sad path should trow type error if one of given numbers is not a real number 8`] = `""" is not a real number"`;
